### ISSUE TITLE
NPC dialogs get offered multiple times if labeled evade

### ISSUE THREAD

**Issue by tehhowch** [endless-sky/endless-sky#2855]
When capturing an NPC labeled `evade` that has a dialog associated with its npc block, the dialog is shown every time,  provided the rest of the NPCs are disabled or destroyed.
The dialog is also not shown automatically if all NPCs are killed.

Contrast with `npc kill`, which only prompts the dialog when all of the npc block have been killed / captured, and not when any ship is captured.

Reference mission is "Wanderers: Sestor: Farpoint Attack 2" - if you defeat the fleet and don't capture one of them, you are not informed about the possibility of not using the bomb. However, if you capture one, you are (immediately) given the dialog.

**Comment by tehhowch** [comment]
Note that for evade, the dialog should also be offered once all the ship are disabled, and not just destroyed. This dialog never appears.

**Comment by tehhowch** [comment]
To the above point about the dialog never appearing when you complete the required flagged action for the npc block (in this case, `evade`): the dialog is shown when the act of recording the corresponding ShipEvent in the map of actions performed (aka past ShipEvents) for these NPCs causes the NPC block to go from failed to succeeded (or not failed to failed). To determine this, the success state is checked before adding the new event to the actions map, and then again after.
The success state is calculated by NPC::HasSucceeded:
```
	// Check what system each ship is in, if there is a requirement that we
	// either evade them, or accompany them. If you are accompanying a ship, it
	// must not be disabled (so that it can land with you). If trying to evade
	// it, disabling it is sufficient (you do not have to kill it).
	if(mustEvade || mustAccompany)
		for(const shared_ptr<Ship> &ship : ships)
		{
			// Special case: if a ship has been captured, it counts as having
			// been evaded.
			auto it = actions.find(ship.get());
			bool isCapturedOrDisabled = ship->IsDisabled();
			if(it != actions.end())
				isCapturedOrDisabled |= (it->second & ShipEvent::CAPTURE);
			bool isHere = (!ship->GetSystem() || ship->GetSystem() == playerSystem);
			if((isHere && !isCapturedOrDisabled) ^ mustAccompany)
				return false;
		}
```
spot the issue?

For `evade`, the npc is in a succeed state if every ship falls into one of these categories: 
1) in a different system than the player
2) destroyed
3) disabled

The issue is that the check of `isCapturedOrDisabled` does not reference the "known" status of the NPC, it references the `actual` status of the NPC, via `ship->IsDisabled()`. So if the event is ShipEvent::DISABLE, on the first pass we compute that oh, nice, success - the NPC is disabled. We then add the ShipEvent::DISABLE to the action record (in NPC::Do). Then, we check again to see if we have succeeded now, which returns the same value - success.

So dialogs predicated on disabling an NPC cannot ever be shown without some changes to `NPC::HasSucceeded`

**Comment by tehhowch** [comment]
The fix to `evade` must also maintain the functionality of `accompany`, e.g. if that NPC is disabled, you cannot succeed until it is repaired. Pretty sure that can be done with an XOR'd ShipEvent::Disabled if the incoming ShipEvent is ShipEvent::ASSIST.

**Comment by tehhowch** [comment]
This should be fixed by #2859
